### ISSUE TITLE
Support the new BCL DateOnly and TimeOnly structs for SQL Server

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#24507]
New DateOnly and TimeOnly structs are being introduced to .NET 6 as alternatives to DateTime (dotnet/runtime#49036). This issue tracks supporting them in SQL Server.

I've opened a separate SqlClient issue for support (https://github.com/dotnet/SqlClient/issues/1009).

**Comment by mattjohnsonpint** [comment]
Please note that the final names for these types are `DateOnly` and `TimeOnly` and that they are now merged into the main branch for the next version of .NET 6 (preview 4 likely).  Please rename the title of this issue accordingly.  Thanks.

**Comment by roji** [comment]
Clearing milestone to discuss doing this in 6.0.0.

**Comment by mattjohnsonpint** [comment]
`DateOnly` and `TimeOnly` should map cleanly to SQL Server's `date` and `time` types.  The tricky part will be to decide what to do with existing behavior that maps them to `DateTime` and `TimeSpan`.  They should continue to function, but the new types should be preferred for things like scaffolding.

**Comment by CleanCodeX** [comment]
As of NET 6 Preview 6 I get a error using EF Core mapping DateOnly properties stating that the DB Provider (SQL Server) does not support this type. Is it supported or not? If not, can HaveConversion bwe a workaround? so far nothing worked (except not using DateOnly).

**Comment by roji** [comment]
@CleanCodeX EF Core support for DateOnly/TimeOnly depends on lower-level support in SqlClient; that's tracked in https://github.com/dotnet/SqlClient/issues/1009.

**Comment by CleanCodeX** [comment]
As already discussed there, it's not supported. What's the common workaround? 
`builder.Properties<DateOnly>().HaveConversion(typeof(DateTime));`
does not do the trick.
If possible at all, do I need to write a custom Converter/Comparer to make it work or is **anything** built-in available that works as a workaround to keep using DateOnly?

**Comment by roji** [comment]
@CleanCodeX there isn't anything built-in for SQL Server, you'll have to implement your own value converter.

**Comment by CleanCodeX** [comment]
> @CleanCodeX there isn't anything built-in for SQL Server, you'll have to implement your own value converter.

So I did.
In case anyone has the same problem:
This seems to do the trick. (at least I get no errors anymore...) If you find any error within, feel free to correct it.

```csharp

      public class YourDbContext : DbContext
      {
            Ctors() {...}

            protected override void ConfigureConventions(ModelConfigurationBuilder builder)
            {
                builder.Properties<DateOnly>().HaveConversion<DateOnlyConverter, DateOnlyComparer>();
            }
      }
              
      /// <summary>
      /// Converts <see cref="DateOnly" /> to <see cref="DateTime"/> and vice versa.
      /// </summary>
      public class DateOnlyConverter : ValueConverter<DateOnly, DateTime>
      {
          /// <summary>
          /// Creates a new instance of this converter.
          /// </summary>
          public DateOnlyConverter() : base(
                  d => d.ToDateTime(TimeOnly.MinValue),
                  d => DateOnly.FromDateTime(d))
          { }
      }
      
      /// <summary>
      /// Compares <see cref="DateOnly" />.
      /// </summary>
      public class DateOnlyComparer : ValueComparer<DateOnly>
      {
          /// <summary>
          /// Creates a new instance of this converter.
          /// </summary>
          public DateOnlyComparer() : base(
              (d1, d2) => d1 == d2 && d1.DayNumber == d2.DayNumber, 
              d => d.GetHashCode())
          {
          }
      }
```

I hope the dev community must not wait until NET 7, that a basic converter like this is added internally so users can use the short version of it:

```csharp

    protected override void ConfigureConventions(ModelConfigurationBuilder builder)
    {
        builder.Properties<DateOnly>().HaveConversion<DateTime>();
        // OR
        builder.Properties<DateOnly>().HaveConversion<DateOnlyToDateTimeConverter>();
    }
```

Thanks for your time. Keep up the good work.

**Comment by akamud** [comment]
> I hope the dev community must not wait until NET 7, that a basic converter like this is added internally so users can use the short version of it:

It isn't "a basic converter" because you are still using datetime in the database. It is a great workaround for anyone who want to just map to `DateOnly`/`TimeOnly` types, but I believe your solution won't easily understand querying on DateOnly/TimeOnly types, mapping from/to migrations, do scaffolding, etc. The official support for SQL Server should probably use `date` and `time` on the database [when appropriate](https://docs.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql?view=sql-server-ver15).

Just being fare with the team working on this ;)

**Comment by roji** [comment]
> It isn't just "a basic converter" because you are still using datetime in the database.

That's true, though it's possible to map DateTime to SQL Server `date` or `time` columns ([using the Fluent API or Data Annotations](https://docs.microsoft.com/en-us/ef/core/modeling/entity-properties?tabs=data-annotations%2Cwithout-nrt#column-data-types)). This would be a bit convoluted - use EF Core value converters to go from DateOnly to DateTime, and then tell SqlClient to write it to SQL Server `date` - but it would work.

Another thing missing from the above is translations from properties/methods on DateOnly/TimeOnly; so this would only take care of reading and writing the values.

**Comment by CleanCodeX** [comment]
Yeah of course the the properties needed to be marked as date / time sql type (or with fluent api). I thought that would be clear then. My mistake if it wasn't.

**Comment by davidhenley** [comment]
What is the status on this with RC1? Do we still need to write custom converters?

**Comment by ErikEJ** [comment]
DateOnly and TimeOnly support for SQL Server is not ready yet, earliest EF Core 7 is my guess.

**Comment by jlorek** [comment]
> > @CleanCodeX there isn't anything built-in for SQL Server, you'll have to implement your own value converter.
> 
> So I did. In case anyone has the same problem: This seems to do the trick. (at least I get no errors anymore...) If you find any error within, feel free to correct it.
> 
> Edit: missing call to HaveColumnType added Edit 2: as of RC2 the Converter for nullable DateOnly isn't recognized anymore.
> 
> ```cs
>       public class YourDbContext : DbContext
>       {
>             Ctors() {...}
> 
>         protected override void ConfigureConventions(ModelConfigurationBuilder builder)
>         {
>             builder.Properties<DateOnly>()
>                 .HaveConversion<DateOnlyConverter, DateOnlyComparer>()
>                 .HaveColumnType("date");
> 
>             builder.Properties<DateOnly?>()
>                 .HaveConversion<NullableDateOnlyConverter, NullableDateOnlyComparer>()
>                 .HaveColumnType("date");
>         }
>       }
>               
>       /// <summary>
>       /// Converts <see cref="DateOnly" /> to <see cref="DateTime"/> and vice versa.
>       /// </summary>
>       public class DateOnlyConverter : ValueConverter<DateOnly, DateTime>
>       {
>           /// <summary>
>           /// Creates a new instance of this converter.
>           /// </summary>
>           public DateOnlyConverter() : base(
>                   d => d.ToDateTime(TimeOnly.MinValue),
>                   d => DateOnly.FromDateTime(d))
>           { }
>       }
>       
>       /// <summary>
>       /// Compares <see cref="DateOnly" />.
>       /// </summary>
>       public class DateOnlyComparer : ValueComparer<DateOnly>
>       {
>           /// <summary>
>           /// Creates a new instance of this converter.
>           /// </summary>
>           public DateOnlyComparer() : base(
>               (d1, d2) => d1 == d2 && d1.DayNumber == d2.DayNumber, 
>               d => d.GetHashCode())
>           {
>           }
>       }
> 
>       /// <summary>
>       /// Converts <see cref="DateOnly?" /> to <see cref="DateTime?"/> and vice versa.
>       /// </summary>
>       public class NullableDateOnlyConverter : ValueConverter<DateOnly?, DateTime?>
>       {
>         /// <summary>
>         /// Creates a new instance of this converter.
>         /// </summary>
>         public NullableDateOnlyConverter() : base(
>             d => d == null 
>                 ? null 
>                 : new DateTime?(d.Value.ToDateTime(TimeOnly.MinValue)),
>             d => d == null 
>                 ? null 
>                 : new DateOnly?(DateOnly.FromDateTime(d.Value)))
>         { }
>     }
> 
>     /// <summary>
>     /// Compares <see cref="DateOnly?" />.
>     /// </summary>
>     public class NullableDateOnlyComparer : ValueComparer<DateOnly?>
>     {
>         /// <summary>
>         /// Creates a new instance of this converter.
>         /// </summary>
>         public NullableDateOnlyComparer() : base(
>             (d1, d2) => d1 == d2 && d1.GetValueOrDefault().DayNumber == d2.GetValueOrDefault().DayNumber,
>             d => d.GetHashCode())
>         {
>         }
>     }
> ```
> 
> I hope the dev community must not wait until NET 7, that a basic converter like this is added internally so users can use the short version of it:
> 
> ```cs
>     protected override void ConfigureConventions(ModelConfigurationBuilder builder)
>     {
>         builder.Properties<DateOnly>().HaveConversion<DateTime>();
>         // OR
>         builder.Properties<DateOnly>().HaveConversion<DateOnlyToDateTimeConverter>();
>     }
> ```
> 
> Thanks for your time. Keep up the good work.

Thanks for the Code, it's really helping me out! But are these ValueComparer Implementations really necessary, since DateOnly already implements IComparable itself?

**Comment by mikelyncheski** [comment]
Thanks for the sample.

But seriously what good are these types without out-of-the-box JSON and EF SQL Server support?

**Comment by roji** [comment]
@mikelyncheski it takes time for new features to propagate throughout the entire stack - that's just the way technology works. Support for DateOnly and TimeOnly did get in for many other database providers (PostgreSQL, SQLite, MySQL) so that's definitely a positive thing; but in other cases there's simply not enough resources to do everything right away.

**Comment by RamType0** [comment]
Why is this still marked as blocking?
#26429 is solved, and [it was already implemented for Npgsql .](https://github.com/npgsql/npgsql/issues/3617)
Is there any SQL Server specific problems?

**Comment by AndriySvyryd** [comment]
@RamType0 https://github.com/dotnet/SqlClient/issues/1009

**Comment by cdrfenix** [comment]
Will this be on EFCore 7 seeing that this is already completed in SqlClient?

https://github.com/dotnet/SqlClient/pull/1813

**Comment by roji** [comment]
No - a SqlClient version hasn't even been released with this functionality. I plan to work on this for 8.0.

**Comment by ErikEJ** [comment]
@roji Could it potentially be retrofitted for EF Core6/7 via a third party package?

**Comment by roji** [comment]
@ErikEJ yeah, that's possible - the same type mapping/query translation code that would go into the SQL Server provider for 8.0 could go into a plugin for older versions, where the plugin would take a dependency on the new SqlClient. Let's revisit this once SqlClient is released with this support.

**Comment by davidbuckleyni** [comment]
Have these been removed in .net 7 and 2019 I get

[]( at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The 'DateOnly?' property 'Supplier.LastModifiedDate' could not be mapped to the database type 'LastModifiedDate' because the database provider does not support mapping 'DateOnly?' properties to 'LastModifiedDate' columns. Consider mapping to a different database type or converting the property value to a type supported by the database using a value converter. See https://aka.ms/efcore-docs-value-converters for more information. Alternately, exclude the property from the model using the '[NotMapped]' attribute or by using 'EntityTypeBuilder.Ignore' in 'OnModelCreating'.)

**Comment by roji** [comment]
@davidbuckleyni DateOnly and TimeOnly aren't yet supported when using the SQL Server provider, that's what this issue tracks.

**Comment by ErikEJ** [comment]
I have just published a preview package that adds support for DateOnly and TimeOnly with EF Core 6 and 7.

Feel free to give it a try and provide feedback in the github repo.

https://www.nuget.org/packages/ErikEJ.EntityFrameworkCore.SqlServer.DateOnlyTimeOnly/6.0.0-preview.1#readme-body-tab

**Comment by WellspringCS** [comment]
@ErikEJ your timing couldn't have been more perfect for me. Giving it a whirl now.

**Comment by SimonCropp** [comment]
doesnt the existence of ErikEJ.EntityFrameworkCore.SqlServer.DateOnlyTimeOnly imply that it should be possible for DateOnly and TimeOnly support to be added into the main EF codebase without waiting for a SqlClient update? @roji

**Comment by ErikEJ** [comment]
@SimonCropp no. My extension depends on SqlClient 5.1 preview. If a LTS version is released before EF Core 8 it will be possible.

**Comment by roji** [comment]
@SimonCropp yeah, as @ErikEJ wrote, an LTS version of SqlClient with DateOnly/TimeOnly support is expected to land before EF Core 8.0 is released (in fact, it may quite soon). Once that happens, we can add native support in the EF SQL Server provider.

**Comment by SimonCropp** [comment]
@roji @ErikEJ thanks for the clarification

**Comment by elachlan** [comment]
We ran into this and we've had to use a work around. Many thanks Erik for your work unblocking this.

**Comment by josegerar** [comment]
I recommend that you review the AspNetCoreDateAndTimeOnly nuget library of my own creation that supports these data types

**Comment by Lonli-Lokli** [comment]
Is there any way to add this support to scaffold as well? Just to not override types manually after generation with db-first

**Comment by ErikEJ** [comment]
@Lonli-Lokli My package should support Scaffolding as well. 

But I will remind myself to add to EF Core Power Tools actually! https://github.com/ErikEJ/EFCorePowerTools/issues/1642

**Comment by davidbuckleyni** [comment]
> @davidbuckleyni DateOnly and TimeOnly aren't yet supported when using the SQL Server provider, that's what this issue tracks.

I new this ages ago dont no why only replying now

**Comment by djfuzzy** [comment]
I recently updated a project to .net 8, including EF Core 8. I noticed the support for `DateOnly` was added so I updated some of my entities to use that instead of `DateTime`. I added the migrations and updated my database succesfully. However, when I go to run, I am getting the following exception:

```
Exception has occurred: CLR/System.InvalidOperationException
An unhandled exception of type 'System.InvalidOperationException' occurred in Microsoft.EntityFrameworkCore.dll: 'The property 'Account.EndDate' could not be mapped because it is of type 'DateOnly?', which is not a supported primitive type or a valid entity type. Either explicitly map this property, or ignore it using the '[NotMapped]' attribute or by using 'EntityTypeBuilder.Ignore' in 'OnModelCreating'.'
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.ValidatePropertyMapping(IModel model, IDiagnosticsLogger`1 logger)
```

Anyone have any insight into why this is happening?
